code/esp_code/tcp_client.c

```diff
@@ -0,0 +1,217 @@
+}
+
+// buffers for spi
+DMA_ATTR char sendbuf[129] = "";
+DMA_ATTR char recvbuf[129] = "";
+
+static void tcp_client_task(void *pvParameters) {
+    // char rx_buffer[128];
+    char addr_str[128];
+    int addr_family;
+    int ip_protocol;
+
+    while (1) {
+
+#ifdef CONFIG_EXAMPLE_IPV4
+        struct sockaddr_in dest_addr;
+        dest_addr.sin_addr.s_addr = inet_addr(HOST_IP_ADDR);
+        dest_addr.sin_family = AF_INET;
+        dest_addr.sin_port = htons(PORT);
+        addr_family = AF_INET;
+        ip_protocol = IPPROTO_IP;
+        inet_ntoa_r(dest_addr.sin_addr, addr_str, sizeof(addr_str) - 1);
+#else // IPV6
+        struct sockaddr_in6 dest_addr;
+        inet6_aton(HOST_IP_ADDR, &dest_addr.sin6_addr);
+        dest_addr.sin6_family = AF_INET6;
+        dest_addr.sin6_port = htons(PORT);
+        addr_family = AF_INET6;
+        ip_protocol = IPPROTO_IPV6;
+        inet6_ntoa_r(dest_addr.sin6_addr, addr_str, sizeof(addr_str) - 1);
+#endif
+
+        int sock = socket(addr_family, SOCK_STREAM, ip_protocol);
+        if (sock < 0) {
+            ESP_LOGE(TAG, "Unable to create socket: errno %d", errno);
+            break;
+        }
+        ESP_LOGI(TAG, "Socket created, connecting to %s:%d", HOST_IP_ADDR,
+                 PORT);
+
+        int err =
+            connect(sock, (struct sockaddr *)&dest_addr, sizeof(dest_addr));
+        if (err != 0) {
+            ESP_LOGE(TAG, "Socket unable to connect: errno %d", errno);
+            break;
+        }
+        ESP_LOGI(TAG, "Successfully connected");
+
+        int n = 0;
+        esp_err_t ret;
+
+        // Configuration for the SPI bus
+        spi_bus_config_t buscfg = {.mosi_io_num = GPIO_MOSI,
+                                   .miso_io_num = GPIO_MISO,
+                                   .sclk_io_num = GPIO_SCLK};
+
+        // Configuration for the SPI slave interface
+        spi_slave_interface_config_t slvcfg = {
+            .mode = 1,
+            .spics_io_num = GPIO_CS,
+            .queue_size = 3,
+            .flags = 0,
+            .post_setup_cb = my_post_setup_cb,
+            .post_trans_cb = my_post_trans_cb};
+
+        // Configuration for the handshake line
+        gpio_config_t io_conf = {.intr_type = GPIO_INTR_DISABLE,
+                                 .mode = GPIO_MODE_OUTPUT,
+                                 .pin_bit_mask = (1 << GPIO_HANDSHAKE)};
+
+        // Configure handshake line as output
+        gpio_config(&io_conf);
+        // Enable pull-ups on SPI lines so we don't detect rogue pulses when no
+        // master is connected.
+        gpio_set_pull_mode(GPIO_MOSI, GPIO_PULLUP_ONLY);
+        gpio_set_pull_mode(GPIO_SCLK, GPIO_PULLUP_ONLY);
+        gpio_set_pull_mode(GPIO_CS, GPIO_PULLUP_ONLY);
+
+        // Initialize SPI slave interface
+        ret = spi_slave_initialize(HSPI_HOST, &buscfg, &slvcfg, 1);
+        assert(ret == ESP_OK);
+
+        memset(recvbuf, 0, 33);
+        spi_slave_transaction_t t;
```

Comment:
You can replace the `memset` with `spi_slave_transaction_t t = { 0 }`. Note: this is different from C++!